vm/src/builtins/code.rs

```diff
@@ -74,16 +75,71 @@ impl Constant for PyConstant {
     }
 }
 
-pub(crate) struct PyObjBag<'a>(pub &'a PyContext);
+pub(crate) struct PyObjBag<'a>(pub &'a VirtualMachine);
 
 impl ConstantBag for PyObjBag<'_> {
     type Constant = PyConstant;
     fn make_constant(&self, constant: bytecode::ConstantData) -> Self::Constant {
-        PyConstant(self.0.unwrap_constant(constant))
+        let vm = self.0;
+        let ctx = &vm.ctx;
+        let obj = match constant {
+            bytecode::ConstantData::Integer { value } => ctx.new_int(value),
+            bytecode::ConstantData::Float { value } => ctx.new_float(value),
+            bytecode::ConstantData::Complex { value } => ctx.new_complex(value),
+            bytecode::ConstantData::Str { value } if value.len() <= 20 => {
+                vm.intern_string(value).into_object()
+            }
+            bytecode::ConstantData::Str { value } => vm.ctx.new_str(value),
+            bytecode::ConstantData::Bytes { value } => ctx.new_bytes(value.to_vec()),
+            bytecode::ConstantData::Boolean { value } => ctx.new_bool(value),
+            bytecode::ConstantData::Code { code } => {
+                ctx.new_code_object(code.map_bag(self)).into_object()
+            }
+            bytecode::ConstantData::Tuple { elements } => {
+                let elements = elements
+                    .into_iter()
+                    .map(|constant| self.make_constant(constant).0)
+                    .collect();
+                ctx.new_tuple(elements)
+            }
+            bytecode::ConstantData::None => ctx.none(),
+            bytecode::ConstantData::Ellipsis => ctx.ellipsis(),
+        };
+        PyConstant(obj)
     }
     fn make_constant_borrowed<C: Constant>(&self, constant: BorrowedConstant<C>) -> Self::Constant {
-        // TODO: check if the constant is a string and try interning it without cloning
-        self.make_constant(constant.into_data())
+        let vm = self.0;
+        let ctx = &vm.ctx;
+        let obj = match constant {
+            bytecode::BorrowedConstant::Integer { value } => ctx.new_bigint(value),
+            bytecode::BorrowedConstant::Float { value } => ctx.new_float(value),
+            bytecode::BorrowedConstant::Complex { value } => ctx.new_complex(value),
+            bytecode::BorrowedConstant::Str { value } if value.len() <= 20 => {
+                vm.intern_string(value).into_object()
+            }
+            bytecode::BorrowedConstant::Str { value } => vm.ctx.new_str(value),
```

Comment:
Does constant interning also affect by string length?

Reply:
Oh, I think I confused it with [this](http://guilload.com/python-string-interning/#avoiding-large-pyc-files), I think it would be interned still.